### DISCUSSION TITLE
udptunnel to route wireguard

### DISCUSSION THREAD

**Discussion by RogerHartmann** [ish-app/ish#1201] [Q&A]
Hello ish Team,

I am trying around to compile udptunnel on ish with the goal to make wireguard work over TCP only connections when the router blocks outgoing UDP traffic on the network. I know there are other ways like OpenVPN, I just wonder if I can make this work instead:

iPad <-> Wireguard client UDP <-> ish udptunnel <-> TCP to wireguard server <-> udptunnel <-> Wireguard server UDP

Just now I tried to compile udptunnel on my ipad in ish but it fails with this message:

host2ip.c:7:10: fatal error: rpcsvc/ypclnt.h: No such file or directory
    7 | #include <rpcsvc/ypclnt.h>   /* YP */

searching the web said I shall install some rpc lib but I could not get this going.

Any suggestion here?
thanks a lot, Roger

**Comment by RogerHartmann** [comment]
Some little progress... after running apk add libnsl-dev the compilation goes further and stops like this:

gcc  -g -O2 -Wall  -o udptunnel  udptunnel.o host2ip.o  
/usr/lib/gcc/i586-alpine-linux-musl/9.3.0/../../../../i586-alpine-linux-musl/bin/ld: host2ip.o: in function `host2ip':
/root/udptunnel/host2ip.c:47: undefined reference to `yp_match'
/usr/lib/gcc/i586-alpine-linux-musl/9.3.0/../../../../i586-alpine-linux-musl/bin/ld: /root/udptunnel/host2ip.c:46: undefined reference to `yp_get_default_domain'
collect2: error: ld returned 1 exit status
make: *** [Makefile:165: udptunnel] Error 1

so further investigating....

**Reply by tbodt** [reply]
Seems you need a -l flag for linking with the right library

**Comment by RogerHartmann** [comment]
thanks @tbodt . I could not get the lib work, but I found a workaround my modifying host2ip.c not to use the functions. Then compilation works.

However, there seems some problems with binding to ports in iSH:

```
# ./udptunnel -c openwrt/80 127.0.0.1/8888 -v
setup_udp_recv: setsockopt(SO_REUSEPORT): Invalid argument
```
to be honest, I have no idea how to maneuver around this one.
By the way, also nc on iSH  does not allow me to use UDP - maybe related, maybe completely independent.
```
# nc -v -u -l -p 12345
listening on [::]:12345 ...
nc: recvfrom: Invalid argument
```

**Comment by RogerHartmann** [comment]
I will look into solving this in python on iSH. Maybe that will be easier. Should not be too much trouble to write a simple UDP/TCP bridge

**Comment by mikeggh** [comment]
I found a udp redirect which has been working for me with wireguard for awhile..
https://github.com/danpodeanu/udp-redirect/blob/main/udp-redirect.c

**Comment by RogerHartmann** [comment]
thanks for the suggestion, @mikeggh . Although iSH is not a use case for me anymore, tunneling wireguard indeed is. Reason: my ISP does not provide an ipv4 address. Some networks, e.g. Hotel Wifis only have ipv4, so I cannot connect directly to my home network. Right now, I use OpenVPN which I route via a virtual server. Nicer would be to use wireguard... let us see if that works

**Comment by RogerHartmann** [comment]
seems like the linked module only supports ipv4... i will check if I can update it to ipv6
